jobmaster/dm/scheduler/scheduler.go

```diff
@@ -0,0 +1,90 @@
+package scheduler
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/tiflow/dm/pkg/log"
+)
+
+const (
+	checkInterval = 50 * time.Millisecond
+)
+
+type Scheduler interface {
+	Schedule(ctx context.Context) error
+}
+
+// DefaultScheduler defines the template method to schedule periodically.
+type DefaultScheduler struct {
+	Scheduler
+
+	mu            sync.RWMutex
+	lastCheckTime time.Time
+	nextCheckTime time.Time
+
+	normalInterval time.Duration
+	errorInterval  time.Duration
+}
+
+func NewDefaultScheduler(normalInterval, errorInterval time.Duration) *DefaultScheduler {
+	defaultScheduler := &DefaultScheduler{
+		normalInterval: normalInterval,
+		errorInterval:  errorInterval,
+	}
+	return defaultScheduler
+}
+
+func (s *DefaultScheduler) setNextCheckTime(t time.Time) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if (t.After(s.lastCheckTime) && t.Before(s.nextCheckTime)) || s.lastCheckTime.Equal(s.nextCheckTime) {
+		s.nextCheckTime = t
+	}
+}
+
+func (s *DefaultScheduler) getNextCheckTime() time.Time {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.nextCheckTime
+}
+
+func (s *DefaultScheduler) advanceCheckTime() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	now := time.Now()
+	s.lastCheckTime = now
+	s.nextCheckTime = now
+}
+
+func (s *DefaultScheduler) Trigger(delay time.Duration) {
+	t := time.Now().Add(delay)
+	s.setNextCheckTime(t)
+}
+
+func (s *DefaultScheduler) Run(ctx context.Context) {
```

Comment:
If this function is called directly by the job master, we can just reuse the `Tick` provided by the framework.

Reply:
It called by job master indeed, But I think separating them reduces coupling.

Reply:
Later it may add some heavy/slow operation, I think tick might be some lightweight operation.

Reply:
I think there're two ways to use `DefaultScheduler`

1.

```
func (j JobMaster) Tick() {
  j.scheduler.Tick()
}
```

2

```
func (j JobMaster) InitImpl() {
  go j.scheduler.Run()
}
```

And I remember @amyangfei says Tick can support some long-time running functions now. I prefer the first way, it's also get rid of implmenting out own tick loop.